### PR TITLE
[@types/lodash] Infer lodash.throttle return type from 'leading' option

### PR DESCRIPTION
**Pull request by dawaltconley**
This sets the `throttle` types to follow the same pattern as `debounce` (which it calls directly), accounting for the different default values (throttle has `leading: true` by default, debounce has `leading: false`).

I'm less familiar with how the `fp` packages are supposed to work, didn't see a way to pass options. I changed the `lodash/fp/throttle` function to return the same value as `lodash/throttle` with default options (`DebounceFuncLeading`).

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  - [lodash.throttle docs](https://lodash.com/docs/4.17.15#throttle)
  - [source](https://github.com/lodash/lodash/blob/main/src/throttle.ts)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

### CODE REVIEW

File: types/lodash/common/function.d.ts [hunk 1]
```diff
@@ -1372,25 +1375,34 @@ declare module "../index" {
          * @param options.trailing Specify invoking on the trailing edge of the timeout.
          * @return Returns the new throttled function.
          */
-        throttle<T extends (...args: any) => any>(func: T, wait?: number, options?: ThrottleSettings): DebouncedFunc<T>;
+        throttle<T extends (...args: any) => any>(func: T, wait: number | undefined, options: ThrottleSettingsNoLeading): DebouncedFunc<T>;
+        throttle<T extends (...args: any) => any>(func: T, wait?: number, options?: ThrottleSettings): DebouncedFuncLeading<T>;
```
Comment (aj-r) [inline types/lodash/common/function.d.ts:1379]: This is a little awkward, since this overload will match if the caller passes a boolean variable (instead of the literal `this` for `leading`. In this case, we don't know if it's leading or not, so ideally the return type should include undefined just in case.

Is there a way to check if the `leading` field is unset? e.g.

```ts
ThrottleSettingsLeadingUnset extends ThrottleSettings {
  leading?: never; // or maybe undefined?
}
```

Then return `DebouncedFuncLeading` only if `leading` is unset or set to true.

File: types/lodash/lodash-tests.ts [hunk 2]
```diff
@@ -3660,21 +3660,27 @@ fp.now(); // $ExpectType number
         leading: true,
         trailing: false,
     };
+    const optionsNoLeading: _.ThrottleSettingsNoLeading = {
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts:3663]: It might be worth writing the options inline, instead of explicitly setting the type here, since most users won't set an explicit type, and that could mess with the overload selection.

File: types/lodash/lodash-tests.ts [hunk 3]
```diff
@@ -3660,21 +3660,27 @@ fp.now(); // $ExpectType number
...
-    _.chain(func).throttle(); // $ExpectType FunctionChain<DebouncedFunc<(a: number, b: string) => boolean>>
-    _.chain(func).throttle(42); // $ExpectType FunctionChain<DebouncedFunc<(a: number, b: string) => boolean>>
-    _.chain(func).throttle(42, options); // $ExpectType FunctionChain<DebouncedFunc<(a: number, b: string) => boolean>>
-
-    fp.throttle(42, func); // $ExpectType DebouncedFunc<(a: number, b: string) => boolean>
-    fp.throttle(42)(func); // $ExpectType DebouncedFunc<(a: number, b: string) => boolean>
+    _.throttle(func); // $ExpectType DebouncedFuncLeading<(a: number, b: string) => boolean>
+    _.throttle(func, 42); // $ExpectType DebouncedFuncLeading<(a: number, b: string) => boolean>
+    _.throttle(func, 42, options); // $ExpectType DebouncedFuncLeading<(a: number, b: string) => boolean>
+    _.throttle(func, 42, optionsNoLeading); // $ExpectType DebouncedFunc<(a: number, b: string) => boolean>
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts:3672]: Add a test with an empty object literal (`{}`) for the options, and a test where `leading` has type `boolean` (or `true|false`).

**Review by aj-r — CHANGES_REQUESTED** [review]
Thanks for the PR!

I'm pretty sure it's impossible to pass options to the FP version, so I think you are correct to always return the leading version.

File: types/lodash/common/function.d.ts (hunk 1 above)
Comment (dawaltconley) [reply types/lodash/common/function.d.ts:1379]: Good point! As an extra complication, I just checked and apparently lodash treats an explicit `{ leading: undefined }` the same as `{ leading: false }`. So unset and undefined fields should be treated differently.

This is what I came up with, going to commit with tests but not sure if there's a way to handle this without some kind of union type. `never` didn't seem to work.

```ts
interface ThrottleSettings {
    leading?: boolean | undefined;
    trailing?: boolean | undefined;
}
type ThrottleSettingsLeading = (ThrottleSettings & { leading: true }) | Omit<ThrottleSettings, 'leading'>
interface LoDashStatic {
    throttle<T extends (...args: any) => any>(func: T, wait?: number, options?: ThrottleSettingsLeading): DebouncedFuncLeading<T>;
    throttle<T extends (...args: any) => any>(func: T, wait?: number, options?: ThrottleSettings): DebouncedFunc<T>;
}
```

Comment (aj-r) [reply types/lodash/common/function.d.ts]: Wow good find! In that case it might be nice to discourage people from setting the field to `undefined` since it almost certainly doesn't do what they expect. But I don't think there's a good way to do that (at least not better than what you have already) so I think we'll have to settle for this.

File: types/lodash/lodash-tests.ts [hunk 4]
```diff
@@ -3656,25 +3656,56 @@ fp.now(); // $ExpectType number
...
+        leading: undefined,
+    };
+    const optionsNoLeading: _.ThrottleSettings = {
+        leading: false,
+    };
+
+    const optionsAmbiguous = {
+        leading: true,
     };
+    const boolean: boolean = false;
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts:3674]: nit: you might need `false as boolean`, otherwise TS will automatically narrow the type to `false` which is not what we want.

File: types/lodash/lodash-tests.ts [hunk 5]
```diff
@@ -3656,25 +3656,56 @@ fp.now(); // $ExpectType number
...
+    const maybeUndefined = (function (): true | undefined { return true })();
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts:3675]: Similar to above: a simpler way to do this might be something like:

```ts
const maybeUndefined: true | undefined = true as any;
```

**Review by aj-r — APPROVED** [review]
Looks great, just a minor tweak for readability

File: types/lodash/lodash-tests.ts [hunk 6]
```diff
@@ -3656,25 +3656,56 @@ fp.now(); // $ExpectType number
...
-    fp.throttle(42, func); // $ExpectType DebouncedFunc<(a: number, b: string) => boolean>
-    fp.throttle(42)(func); // $ExpectType DebouncedFunc<(a: number, b: string) => boolean>
+    _.throttle(func); // $ExpectType DebouncedFuncLeading<(a: number, b: string) => boolean>
+    _.throttle(func, 42); // $ExpectType DebouncedFuncLeading<(a: number, b: string) => boolean>
+    _.throttle(func, 42, {}); // $ExpectType DebouncedFuncLeading<(a: number, b: string) => boolean>
+    _.throttle(func, 42, { leading: true, trailing: false }); // $ExpectType DebouncedFuncLeading<(a: number, b: string) => boolean>
+    _.throttle(func, 42, { leading: false }); // $ExpectType DebouncedFunc<(a: number, b: string) => boolean>
+    _.throttle(func, 42, { leading: undefined }); // $ExpectType DebouncedFunc<(a: number, b: string) => boolean>
+    _.throttle(func, 42, { leading: boolean }); // $ExpectType DebouncedFunc<(a: number, b: string) => boolean>
+    _.throttle(func, 42, { leading: maybeUndefined }); // $ExpectType DebouncedFunc<(a: number, b: string) => boolean>
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts:3686]: Also add a comment explaining why we should NOT expect `DebouncedFuncLeading` here, since this is probably surprising to everyone.

**Review by aj-r — CHANGES_REQUESTED** [review]

File: types/lodash/common/function.d.ts (hunk 1 above)
Comment (dawaltconley) [reply types/lodash/common/function.d.ts]: I just pushed a commit to clarify the tests. Had the thought right after that it _is_ possible to prohibit an explicit undefined here, while still making `leading` an optional property.

```ts
interface ThrottleSettings {
    leading: boolean;
    trailing?: boolean | undefined;
}
type ThrottleSettingsLeading = (ThrottleSettings & { leading: true }) | Omit<ThrottleSettings, 'leading'>
interface LoDashStatic {
    throttle<T extends (...args: any) => any>(func: T, wait?: number, options?: ThrottleSettingsLeading): DebouncedFuncLeading<T>;
    throttle<T extends (...args: any) => any>(func: T, wait?: number, options?: ThrottleSettings): DebouncedFunc<T>;
}
```

But I think that's a breaking change and I don't know if it's preferable anyway.

**Review by aj-r — APPROVED** [review]
Looks great!

File: types/lodash/common/function.d.ts (hunk 1 above)
Comment (aj-r) [reply types/lodash/common/function.d.ts]: Pretty much every type change is a breaking change, due to the way types work. But I think that change still doesn't completely prevent explicit undefined, so it's debatable whether it's worth the potential breakage. (It prevents explicit undefined in object literals but not prebuilt objects. Maybe better than nothing?)

**Review by aj-r — APPROVED** [review]
I'm not sure why typescript-bot thinks there are no reviews...

**Review by sandersn — APPROVED** [review]

**Comment by dawaltconley** [comment]
Ready to merge
